cpp/perspective/src/cpp/computed_expression.cpp

```diff
@@ -62,17 +62,15 @@ t_computed_expression_parser::LOWER_VALIDATOR_FN = computed_function::lower(null
 computed_function::length
 t_computed_expression_parser::LENGTH_VALIDATOR_FN = computed_function::length(nullptr);
 
-// Register functions in a centralized macro instead of copy-pasting the same
-// lines between compute/recompute
 #define REGISTER_COMPUTE_FUNCTIONS()                                                        \
-    computed_function::day_of_week day_of_week_fn = computed_function::day_of_week(m_expression_vocab);         \
-    computed_function::month_of_year month_of_year_fn = computed_function::month_of_year(m_expression_vocab);   \
-    computed_function::intern intern_fn = computed_function::intern(m_expression_vocab);    \
-    computed_function::concat concat_fn = computed_function::concat(m_expression_vocab);    \
-    computed_function::order order_fn = computed_function::order(m_expression_vocab);       \
-    computed_function::upper upper_fn = computed_function::upper(m_expression_vocab);       \
-    computed_function::lower lower_fn = computed_function::lower(m_expression_vocab);       \
-    computed_function::length length_fn = computed_function::length(m_expression_vocab);    \
+    computed_function::day_of_week day_of_week_fn = computed_function::day_of_week(vocab);         \
+    computed_function::month_of_year month_of_year_fn = computed_function::month_of_year(vocab);   \
+    computed_function::intern intern_fn = computed_function::intern(vocab);    \
+    computed_function::concat concat_fn = computed_function::concat(vocab);    \
+    computed_function::order order_fn = computed_function::order(vocab);       \
+    computed_function::upper upper_fn = computed_function::upper(vocab);       \
+    computed_function::lower lower_fn = computed_function::lower(vocab);       \
+    computed_function::length length_fn = computed_function::length(vocab);    \
```

Comment:
Any reason this needs to be a macro and not a function?  The scope-capture of `vocab` specifically is something I'd like to stylistically avoid if possible. (Looks also like `recompute` is deprecated?)

Reply:
I don't think this macro additionally captures the vocab - it is basically just inlining this code into where the macro is invoked, and the macro doesn't take `vocab` as a parameter. AFAIK since it's just a text replacement macro, it shouldn't capture `vocab` or have any notion of what `vocab` is during the preprocessor step. I think leaving it as a macro works well for readability in the `compute` function.

Reply:
[That's what macro variable capture means](https://en.wikipedia.org/wiki/Hygienic_macro#The_hygiene_problem).  There's little risk of `vocab` being mis-used here as you mention the macro won't compile, but for complexity analysis purposes I think its best if we avoid variable capture (in this case with a macro argument) and try where possible to force scope (with `{}`) in our macros.